resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala

```diff
@@ -82,17 +82,36 @@ private[spark] class Client(
   def run(): Unit = {
     validateNoDuplicateFileNames(sparkJars)
     validateNoDuplicateFileNames(sparkFiles)
-    val parsedCustomLabels = ConfigurationUtils.parseKeyValuePairs(
+
+    val parsedCustomLabelsDeprecated = ConfigurationUtils.parseKeyValuePairs(
         customLabels, KUBERNETES_DRIVER_LABELS.key, "labels")
-    require(!parsedCustomLabels.contains(SPARK_APP_ID_LABEL), s"Label with key " +
-      s" $SPARK_APP_ID_LABEL is not allowed as it is reserved for Spark bookkeeping" +
-      s" operations.")
-    val parsedCustomAnnotations = ConfigurationUtils.parseKeyValuePairs(
-      customAnnotations, KUBERNETES_DRIVER_ANNOTATIONS.key, "annotations")
-    require(!parsedCustomAnnotations.contains(SPARK_APP_NAME_ANNOTATION), s"Annotation with key" +
-      s" $SPARK_APP_NAME_ANNOTATION is not allowed as it is reserved for Spark bookkeeping" +
-      s" operations.")
-    val allLabels = parsedCustomLabels ++ Map(
+    // Remark: getAllWithPrefix strips out the prefix in the returned array.
+    val customLabelsFromPrefixConf = sparkConf.getAllWithPrefix(KUBERNETES_DRIVER_LABEL_PREFIX)
+    val allCustomLabels = parsedCustomLabelsDeprecated.toSeq ++ customLabelsFromPrefixConf
+    allCustomLabels.groupBy(_._1).foreach {
+      case (key, values) =>
+        require(values.size == 1,
+            s"Cannot have multiple values for a label key, got key $key with values $values")
+    }
+    require(!allCustomLabels.map(_._1).contains(SPARK_APP_ID_LABEL), s"Label with key " +
+        s" $SPARK_APP_ID_LABEL is not allowed as it is reserved for Spark bookkeeping" +
+        s" operations.")
+
+    val parsedCustomAnnotationsDeprecated = ConfigurationUtils.parseKeyValuePairs(
+      customAnnotations, KUBERNETES_DRIVER_ANNOTATIONS.key, "annotations").toSeq
+    val customAnnotationsFromPrefixConf =
+        sparkConf.getAllWithPrefix(KUBERNETES_DRIVER_ANNOTATION_PREFIX)
+    val allCustomAnnotations = parsedCustomAnnotationsDeprecated ++ customAnnotationsFromPrefixConf
+    allCustomAnnotations.groupBy(_._1).foreach {
+      case (key, values) =>
+        require(values.size == 1,
+            s"Cannot have multiple values for the same annotation key, got key $key" +
+            s"with values $values")
+    }
+    require(!allCustomAnnotations.map(_._1).contains(SPARK_APP_NAME_ANNOTATION),
+        s"Annotation with key $SPARK_APP_NAME_ANNOTATION is not allowed as it is reserved for" +
+        s" Spark bookkeeping operations.")
+    val allLabels = allCustomLabels.toMap ++ Map(
```

Comment:
This is getting long -- we should extract out a method that produces both `allLabels` and `allAnnotations` (named as `driverLabels` and `driverAnnotations`)